### PR TITLE
gitserver: grpc: add new ChangedFiles method to gitserver client

### PR DESCRIPTION
**Pull request by ggilmore**
Part of https://github.com/sourcegraph/sourcegraph/issues/60654

This PR builds on https://github.com/sourcegraph/sourcegraph/pull/62216, and creates a new ChangedFiles method in the gitserver client that calls the new gRPC method that was introduced in https://github.com/sourcegraph/sourcegraph/pull/62216.

Note: This PR doesn't change / remove any existing callers of DiffSymbols. I made those changes in future stacked PRs to ensure that they're easy to review. 

## Test plan

New Unit tests

### CODE REVIEW

**Comment by ggilmore** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/62354?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#62355** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/62355?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#62354** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/62354?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#62262** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/62262?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#62252** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/62252?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#62217** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/62217?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#62216** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/62216?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#62212** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/62212?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#62302** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/62302?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#62173** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/62173?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#62169** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/62169?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#62165** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/62165?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#62158** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/62158?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @ggilmore and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: internal/gitserver/commands.go [hunk 1]
```diff
@@ -20,11 +20,6 @@ import (
 	"github.com/go-git/go-git/v5/plumbing/format/config"
 	"github.com/golang/groupcache/lru"
 	"github.com/sourcegraph/go-diff/diff"
-	"go.opentelemetry.io/otel/attribute"
-	"google.golang.org/grpc/codes"
-	"google.golang.org/grpc/status"
-	"google.golang.org/protobuf/types/known/timestamppb"
```
Comment (eseliger) [inline internal/gitserver/commands.go:26]: this reordering of dependencies seems wrong, and an unnecessary diff (sorry for being the nitpick police here 😬) 
goimports should be able to sort them correctly with the `-local` flag I believe. This should be configured correctly in vscode, maybe that config can be adopted for Goland

File: internal/gitserver/commands.go [hunk 2]
```diff
@@ -346,6 +345,50 @@ func parseTimestamp(timestamp string) (time.Time, error) {
 // the root commit.
 const DevNullSHA = "4b825dc642cb6eb9a060e54bf8d69288fbee4904"
 
+func (c *clientImplementor) ChangedFiles(ctx context.Context, repo api.RepoName, base, head string) (_ []gitdomain.PathStatus, err error) {
```
Comment (eseliger) [inline internal/gitserver/commands.go:348]: Since this can return 32MB+, I wonder since this is a net-new API if we should return a iterator type here instead, so the client package doesn't accumulate the memory and makes the caller think about doing things with best efficiency. 

Not a blocker though, I've not done that for some APIs either. This one's net-new so maybe a bit easier to adapt?

**Review by eseliger — APPROVED** [review]

File: internal/gitserver/commands.go (hunk 2 above)
Comment (keegancsmith) [reply internal/gitserver/commands.go:348]: Edit: Nice! I see the underlying grpc is using streaming. Nice work fellas! removing my original response. Indeed an iterator return here wouldn't be too cumbersome for call sites. I think that would make the work by geoffrey refactoring stuff more tricky though. Maybe to ease the introduction you make the API iterator based, but provide a helper to aggregate for all the call sites to use at first?

**Review by keegancsmith — APPROVED** [review]

File: internal/gitserver/commands.go (hunk 2 above)
Comment (eseliger) [reply internal/gitserver/commands.go:348]: Traversing an iterator and adding it all to a list is a bit verbose but super straightforward in go, the less helpers we provide the less we encourage people to use patterns that potentially use a lot of memory 😬 
But either way is fine with me

**Comment by ggilmore** [comment]
@sourcegraph/source Can you please take another look at this PR (and all the upstack PRs that use it)? I refactored the approach here to use iterators.

**Review by eseliger — APPROVED** [review]

**Comment by ggilmore** [comment]
### Merge activity

* **May 7, 1:45 PM EDT**: [Graphite](https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/62354) rebased this pull request after merging its parent, because this pull request is set to merge when ready.
